### PR TITLE
Support named TaskHub and ConnectionName on the input binding

### PR DESCRIPTION
**Pull request by jviau**
**Depends on #2366** - you will see a bunch of duplicate changes here. They are separate PRs so that this one can eventually be reviewed in a more narrow scope (when 2366 is completed).

Draft PR to get some eyes on this. Will be filling out the PR description properly soon.

### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [ ] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [x] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`

### CODE REVIEW

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs [hunk 1]
```diff
@@ -91,30 +92,32 @@ public ValueTask<ConversionResult> ConvertAsync(ConverterContext converterContex
...
+                    new InvalidOperationException($"Expected the Durable Task WebJobs SDK extension to send a string payload for {nameof(DurableClientAttribute)}.")));
             }
 
             try
             {
-                DurableTaskClient client = this.clientProvider.GetClient(inputData?.rpcBaseUrl);
+                DurableClientInputData? inputData = JsonSerializer.Deserialize<DurableClientInputData>(clientConfigText);
+                if (!Uri.TryCreate(inputData?.rpcBaseUrl, UriKind.Absolute, out Uri? endpoint))
+                {
+                    throw new InvalidOperationException("Failed to parse the input binding payload data");
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:104]: Why are you throwing here when in other places we return the exception as a `ConversionResult`?

File: src/Worker.Extensions.DurableTask/FunctionsDurableClientProvider.cs [hunk 2]
```diff
@@ -130,4 +145,57 @@ private void VerifyNotDisposed()
...
+        public async ValueTask DisposeAsync()
+        {
+            try
+            {
+                await this.Client.DisposeAsync();
+                await this.channel.ShutdownAsync();
+            }
+            catch
+            {
+                // dispose should not through and unsure how Channel multiple ShutdownAsync() calls behave.
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/FunctionsDurableClientProvider.cs:171]: through --> throw?

File: src/Worker.Extensions.DurableTask/FunctionsDurableClientProvider.net6.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    private static GrpcChannel CreateChannel(ClientKey key)
+    {
+        IReadOnlyDictionary<string, string> headers = key.GetHeaders();
+        if (headers.Count == 0)
+        {
+            return GrpcChannel.ForAddress(key.Address);
+        }
+
+
+        HttpClient httpClient = new();
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/FunctionsDurableClientProvider.net6.cs:22]: How often are we creating these?

**Review by cgillum — COMMENTED** [review]
Overall looks good. I think there are some lifetime things related to channels that I don't quite understand/remember, so I asked a related question below.

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs (hunk 1 above)
Comment (jviau) [reply src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:104]: We will catch and return as `ConversionResult`. I can switch to just returning `ConversionResult`.

File: src/Worker.Extensions.DurableTask/FunctionsDurableClientProvider.net6.cs (hunk 3 above)
Comment (jviau) [reply src/Worker.Extensions.DurableTask/FunctionsDurableClientProvider.net6.cs:22]: Only once per unique `(endpoint, taskHub, connectionName` tuple. In a typical app, only one for the lifetime of the app.

**Review by cgillum — APPROVED** [review]
LGTM!

**Comment by jviau** [comment]
> I think there are some lifetime things related to channels that I don't quite understand/remember, so I asked a related question below.

This provider is `Singleton` lifetime, so really nothing is disposed for the lifetime of the worker. I used caching to keep allocations of new clients to an absolute minimum, but also still wanted to make sure the disposal behavior was 'correct' (even though it typically is never used).

**Pull request merged**